### PR TITLE
[2204] Add styled string for tree item labels

### LINKED ISSUE #2204: Add styled string for tree item labels

### PR DESCRIPTION
**Pull request by mcharfadi**
Bug: https://github.com/eclipse-sirius/sirius-components/issues/2204

### CODE REVIEW

File: packages/forms/backend/sirius-components-widget-reference/src/main/java/org/eclipse/sirius/components/widget/reference/browser/ModelBrowserDescriptionProvider.java [hunk 1]
```diff
@@ -43,6 +43,7 @@
 import org.eclipse.sirius.components.representations.IStatus;
 import org.eclipse.sirius.components.representations.VariableManager;
 import org.eclipse.sirius.components.trees.description.TreeDescription;
+import org.eclipse.sirius.components.trees.labels.*;
```
Comment (sbegaudeau) [inline packages/forms/backend/sirius-components-widget-reference/src/main/java/org/eclipse/sirius/components/widget/reference/browser/ModelBrowserDescriptionProvider.java]: Don't use star imports

File: packages/core/frontend/sirius-components-core/src/label/StyledLabel.tsx [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+import Typography from '@material-ui/core/Typography';
+import { makeStyles } from '@material-ui/core/styles';
+import { useState } from 'react';
+import { GQLStyledString, StyledLabelInputProps, StyledLabelInputState } from './StyledLabel.type';
+
+const useTreeItemStyle = makeStyles((theme) => ({
+  selected: {
+    stroke: '#66808a',
+    fill: '#66808a',
+    backgroundColor: 'var(--blue-lagoon-lighten-90)',
```
Comment (sbegaudeau) [inline packages/core/frontend/sirius-components-core/src/label/StyledLabel.tsx]: Don't use either "magic colors" nor colors from outdated style `(var--xxx)`

File: packages/core/frontend/sirius-components-core/src/label/StyledLabel.type.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+export interface GQLStyledString {
+  fragments: GQLFragment[];
+}
+
+export interface GQLFragment {
+  text: string;
+  style: GQLStyle;
+}
+
+export interface GQLStyle {
```
Comment (sbegaudeau) [inline packages/core/frontend/sirius-components-core/src/label/StyledLabel.type.ts]: Tons of things could be called Style, please use a more specific name `StyledStringFragment` and `StyledStringFragmentStyle`

File: packages/emf/backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/services/ObjectService.java [hunk 4]
```diff
@@ -125,13 +121,39 @@ private String getIdFromURIFragment(EObject eObject) {
...
+                if (adapter == null) {
+                    //@formatter:off
+                    return Optional.of(this.getLabelEAttribute(eObject).map(eObject::eGet))
+                            .map(Object::toString)
+                            .orElse("");
+                    //@formatter:on
+                }
+            }
+        }
+        return label;
```
Comment (sbegaudeau) [inline packages/emf/backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/services/ObjectService.java]: If we were to keep a `String getLabel(Object object)` method, the only valid content inside of this method would be this:

```
public String getLabel(Object object) {
    return this.getStyledLabel(object).toString();
}

public StyledString getStyledLabel(Object object) {
    // ....
}

```

File: packages/trees/backend/sirius-components-trees/src/main/java/org/eclipse/sirius/components/trees/labels/Style.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+ *     Obeo - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.sirius.components.trees.labels;
+
+/**
+ * Fragment's style.
+ *
+ * @author mcharfadi
+ */
+public record Style(String font, String backgroundColor, String foregroundColor, String strikeoutColor, String underlineColor, String borderColor, boolean isStrikedout, UnderLineStyle underLineStyle, BorderStyle boderStyle) { }
```
Comment (sbegaudeau) [inline packages/trees/backend/sirius-components-trees/src/main/java/org/eclipse/sirius/components/trees/labels/Style.java]: - Use a builder pattern to help create those objects
- Use a more meaningful name (see other comment on the matter)

File: packages/trees/backend/sirius-components-trees/src/main/java/org/eclipse/sirius/components/trees/labels/StyledString.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * A String composed of Fragment that each define a style.
+ *
+ * @author mcharfadi
+ */
+public record StyledString(List<Fragment> fragments) {
+    public String getString() {
```
Comment (sbegaudeau) [inline packages/trees/backend/sirius-components-trees/src/main/java/org/eclipse/sirius/components/trees/labels/StyledString.java]: Override `toString()` instead

File: packages/trees/backend/sirius-components-trees/src/main/java/org/eclipse/sirius/components/trees/labels/UnderLineStyle.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ * @author mcharfadi
+ */
+public enum UnderLineStyle {
+    NONE,
+    SINGLE,
+    DOUBLE,
+    SQUIGGLE,
+    ERROR,
+    LINK,
```
Comment (sbegaudeau) [inline packages/trees/backend/sirius-components-trees/src/main/java/org/eclipse/sirius/components/trees/labels/UnderLineStyle.java]: Use exactly the CSS names

File: packages/forms/backend/sirius-components-widget-reference/src/main/java/org/eclipse/sirius/components/widget/reference/browser/ModelBrowserDescriptionProvider.java [hunk 8]
```diff
@@ -183,7 +184,9 @@ private String getLabel(VariableManager variableManager) {
                 label = this.urlParser.getParameterValues(kind).get(SemanticKindConstants.ENTITY_ARGUMENT).get(0);
             }
         }
-        return label;
+        Style style = new Style("", "", "", "", "", "", false, UnderLineStyle.NONE, BorderStyle.NONE);
```
Comment (sbegaudeau) [inline packages/forms/backend/sirius-components-widget-reference/src/main/java/org/eclipse/sirius/components/widget/reference/browser/ModelBrowserDescriptionProvider.java]: A builder pattern is absolutely require for that

**Review by sbegaudeau — CHANGES_REQUESTED** [review]

File: integration-tests/cypress/e2e/project/edit/color-palette.cy.js [hunk 9]
```diff
@@ -22,21 +22,21 @@ describe('/projects/:projectId/edit - Color Palette', () => {
   it('check the default color palette exist and populated', () => {
     cy.getByTestId('ViewNewModel-toggle').click();
     cy.getByTestId('View-toggle').click();
-    cy.getByTestId('ColorPalette').should('exist');
-    cy.getByTestId('ColorPalette-toggle').click();
-    cy.getByTestId('ColorPalette').should('exist');
+    cy.getByTestId('Color Palette').should('exist');
```
Comment (sbegaudeau) [inline integration-tests/cypress/e2e/project/edit/color-palette.cy.js]: No need to change the data-tested, I don't even know how many tests are relying on them in Sirius Web and in other applications which would have to be updated.

File: integration-tests/cypress/e2e/project/edit/explorer-filterbar.cy.js [hunk 10]
```diff
@@ -32,36 +32,36 @@ describe('/projects/:projectId/edit - Tree filter bar', () => {
 
   it('type text in filter bar', () => {
     cy.getByTestId('flow').dblclick();
-    cy.getByTestId('Robot').dblclick();
-    cy.getByTestId('Central_Unit').dblclick();
-    cy.getByTestId('CaptureSubSystem').dblclick();
-    cy.getByTestId('Wifi').dblclick();
+    cy.getByTestId('System Robot').dblclick();
```
Comment (sbegaudeau) [inline integration-tests/cypress/e2e/project/edit/explorer-filterbar.cy.js]: I won't comment it on every single line but those should not change

**Comment by sbegaudeau** [comment]
With the introduction of the composed object service, we will now be able to move forward by merging this since we will be able to use it for some examples such as the view and domain metamodels

**Comment by frouene** [comment]
Where the file `treeItemLabelStyled.jpg` is used ?

File: packages/core/backend/sirius-components-core/src/main/java/org/eclipse/sirius/components/core/api/ILabelService.java [hunk 11]
```diff
@@ -42,6 +46,11 @@ public String getLabel(Object object) {
             return "";
         }
 
+        @Override
+        public StyledString getStyledLabel(Object object) {
+            return StyledString.of("");
```
Comment (frouene) [inline packages/core/backend/sirius-components-core/src/main/java/org/eclipse/sirius/components/core/api/ILabelService.java:51]: Why don't use the same noop that for `IDefaultLabelService`

**Comment by mcharfadi** [comment]
It should have been used in the changelog, I'll link it back correctly

File: packages/core/frontend/sirius-components-core/src/label/StyledLabel.tsx [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Obeo - initial API and implementation
+ *******************************************************************************/
+import Typography from '@material-ui/core/Typography';
+import { makeStyles } from '@material-ui/core/styles';
+import { GQLStyledString, GQLStyledStringFragmentStyle, StyledLabelInputProps } from './StyledLabel.type';
+
+export const splitText = (label: string, userInput: string | null): string[] => {
```
Comment (frouene) [inline packages/core/frontend/sirius-components-core/src/label/StyledLabel.tsx:18]: does it need to be exported ?

File: packages/core/frontend/sirius-components-core/src/label/StyledLabel.tsx [hunk 13]
```diff
@@ -0,0 +1,176 @@
+/*******************************************************************************
+ * Copyright (c) 2023, 2024 Obeo.
```
Comment (frouene) [inline packages/core/frontend/sirius-components-core/src/label/StyledLabel.tsx]: wrong dates

**Review by frouene — CHANGES_REQUESTED** [review]

File: packages/core/backend/sirius-components-core/src/main/java/org/eclipse/sirius/components/core/api/ILabelService.java (hunk 11 above)
Comment (mcharfadi) [reply packages/core/backend/sirius-components-core/src/main/java/org/eclipse/sirius/components/core/api/ILabelService.java:51]: What do you mean ? The content is the same but they don't share an interface.

File: packages/core/frontend/sirius-components-core/src/label/StyledLabel.tsx (hunk 12 above)
Comment (mcharfadi) [reply packages/core/frontend/sirius-components-core/src/label/StyledLabel.tsx:18]: Yes, it's used in `FilterableSortableList`

Comment (frouene) [reply packages/core/frontend/sirius-components-core/src/label/StyledLabel.tsx:18]: In `FilterableSortableList`, the one used is from `'@eclipse-sirius/sirius-components-trees'`
By the way, `splitText` is duplicated in `sirius-components-trees` and `sirius-components-core`

File: packages/core/backend/sirius-components-core/src/main/java/org/eclipse/sirius/components/core/api/ILabelService.java (hunk 11 above)
Comment (frouene) [reply packages/core/backend/sirius-components-core/src/main/java/org/eclipse/sirius/components/core/api/ILabelService.java:51]: In `IDefaultLabelService` the NoOp impl is `return new StyledString(List.of());`
Why not use the same default impl in `ILabelService` ?
(It's just a small detail to be consistent, it doesn't really matter)

File: packages/core/frontend/sirius-components-core/src/label/StyledLabel.tsx (hunk 12 above)
Comment (mcharfadi) [reply packages/core/frontend/sirius-components-core/src/label/StyledLabel.tsx:18]: You're right there is an inconsistency there.
I think `splitText `should be exported from `sirius-components-core` and `getText `should be renamed to `getTextFromStyledLabel ` and also be exported from `sirius-components-core`. Would that be okay ?

File: packages/core/backend/sirius-components-core/src/main/java/org/eclipse/sirius/components/core/api/ILabelService.java (hunk 11 above)
Comment (mcharfadi) [reply packages/core/backend/sirius-components-core/src/main/java/org/eclipse/sirius/components/core/api/ILabelService.java:51]: Okay I see, I'll make the change then.

File: packages/core/frontend/sirius-components-core/src/label/StyledLabel.tsx (hunk 12 above)
Comment (frouene) [reply packages/core/frontend/sirius-components-core/src/label/StyledLabel.tsx:18]: Looks good to me

**Review by frouene — APPROVED** [review]

File: packages/core/frontend/sirius-components-core/src/label/StyledLabel.type.ts [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+  foregroundColor: string;
+  isStrikedout: boolean;
+  strikeoutColor: string;
+  underlineColor: string;
+  borderColor: string;
+  borderStyle: GQLBorderStyle;
+  underlineStyle: GQLUnderLineStyle;
+}
+
+enum GQLUnderLineStyle {
```
Comment (sbegaudeau) [inline packages/core/frontend/sirius-components-core/src/label/StyledLabel.type.ts]: Never use TypeScript `enum`, it does not do what you think it does, use unions instead

File: packages/core/frontend/sirius-components-core/src/label/StyledLabel.type.ts [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+  NONE,
+  SOLID,
+  DOUBLE,
+  DOTTED,
+  DASHED,
+  WAVY,
+}
+
+enum GQLBorderStyle {
```
Comment (sbegaudeau) [inline packages/core/frontend/sirius-components-core/src/label/StyledLabel.type.ts]: Same comment

File: packages/forms/backend/sirius-components-collaborative-widget-reference/src/main/java/org/eclipse/sirius/components/collaborative/widget/reference/browser/ModelBrowsersDescriptionProvider.java [hunk 16]
```diff
@@ -258,19 +260,25 @@ private String getKind(VariableManager variableManager) {
         return kind;
     }
 
-    private String getLabel(VariableManager variableManager) {
+    private StyledString getLabel(VariableManager variableManager) {
         Object self = variableManager.getVariables().get(VariableManager.SELF);
         String label = "";
-        if (self instanceof Resource resource) {
+        if (self instanceof RepresentationMetadata representationMetadata) {
+            label = representationMetadata.getLabel();
```
Comment (sbegaudeau) [inline packages/forms/backend/sirius-components-collaborative-widget-reference/src/main/java/org/eclipse/sirius/components/collaborative/widget/reference/browser/ModelBrowsersDescriptionProvider.java]: The reference widget does not display representations

File: packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/services/tree/ECoreSampleStyledLabelService.java [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+import org.eclipse.sirius.components.emf.services.LabelFeatureProviderRegistry;
+import org.springframework.stereotype.Service;
+
+/**
+ * Implementation of ILabelServiceDelegate to test styled tree items.
+ *
+ * @author mcharfadi
+ */
+@Service
+public class ECoreSampleStyledLabelService extends DefaultLabelService implements ILabelServiceDelegate {
```
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/services/tree/ECoreSampleStyledLabelService.java]: ECore -> Ecore

File: packages/trees/backend/sirius-components-collaborative-trees/src/main/resources/schema/tree.graphqls [hunk 18]
```diff
@@ -46,6 +46,52 @@ type TreeItem {
...
+    isStrikedout: Boolean
+    underlineStyle: UnderLineStyle
+    borderStyle: BorderStyle
+    font: String
+    backgroundColor: String
+    foregroundColor: String
+    strikeoutColor: String
+    underlineColor: String
+    borderColor: String
+}
```
Comment (sbegaudeau) [inline packages/trees/backend/sirius-components-collaborative-trees/src/main/resources/schema/tree.graphqls]: formatting issue (two spaces vs four spaces)

File: packages/trees/backend/sirius-components-trees/pom.xml [hunk 19]
```diff
@@ -45,6 +45,11 @@
 			<artifactId>sirius-components-representations</artifactId>
 			<version>2024.7.9</version>
 		</dependency>
+		<dependency>
+			<groupId>org.eclipse.sirius</groupId>
+			<artifactId>sirius-components-core</artifactId>
+			<version>2024.7.9</version>
+		</dependency>
```
Comment (sbegaudeau) [inline packages/trees/backend/sirius-components-trees/pom.xml]: sirius-components-core already depends on sirius-components-representations which can thus be removed as a direct dependency

File: packages/trees/frontend/sirius-components-trees/src/views/__tests__/GraphqlTreeEventSubscription.test.ts [hunk 20]
```diff
@@ -40,7 +40,22 @@ const getDocumentSubscription = gql`
     id
     hasChildren
     expanded
-    label
+    label {
+      styledStringFragments {
+        text
+        styledStringFragmentStyle {
+          isStrikedout
```
Comment (sbegaudeau) [inline packages/trees/frontend/sirius-components-trees/src/views/__tests__/GraphqlTreeEventSubscription.test.ts]: I won't make you change it now but I don't think that `isStrikedOut` is valid in English. I think that it should be `isStruckOut`. We will change it later

**Review by sbegaudeau — COMMENTED** [review]
I'll rebase the PR and look to fix the minor issues detected. Once it will be done, I'll test it a little to merge it if I don't encounter any major issue.

File: packages/core/frontend/sirius-components-core/src/label/StyledLabel.type.ts (hunk 14 above)
Comment (sbegaudeau) [reply packages/core/frontend/sirius-components-core/src/label/StyledLabel.type.ts]: Done

File: packages/core/frontend/sirius-components-core/src/label/StyledLabel.type.ts (hunk 15 above)
Comment (sbegaudeau) [reply packages/core/frontend/sirius-components-core/src/label/StyledLabel.type.ts]: Done

File: packages/forms/backend/sirius-components-collaborative-widget-reference/src/main/java/org/eclipse/sirius/components/collaborative/widget/reference/browser/ModelBrowsersDescriptionProvider.java (hunk 16 above)
Comment (sbegaudeau) [reply packages/forms/backend/sirius-components-collaborative-widget-reference/src/main/java/org/eclipse/sirius/components/collaborative/widget/reference/browser/ModelBrowsersDescriptionProvider.java]: Done

File: packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/services/tree/ECoreSampleStyledLabelService.java (hunk 17 above)
Comment (sbegaudeau) [reply packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/services/tree/ECoreSampleStyledLabelService.java]: Done

File: packages/trees/backend/sirius-components-collaborative-trees/src/main/resources/schema/tree.graphqls (hunk 18 above)
Comment (sbegaudeau) [reply packages/trees/backend/sirius-components-collaborative-trees/src/main/resources/schema/tree.graphqls]: There were other formatting issues that I've fixed at the same time

File: packages/trees/backend/sirius-components-trees/pom.xml (hunk 19 above)
Comment (sbegaudeau) [reply packages/trees/backend/sirius-components-trees/pom.xml]: Please try to be more careful with dependencies next time guys, I can't be the only one caring about things like that /cc @mcharfadi @frouene

**Review by sbegaudeau — COMMENTED** [review]
I'll wait for the build to finally merge it, I've tested it quickly and the end result looks nice

**Review by sbegaudeau — APPROVED** [review]

**Pull request merged**